gensim/models/poincare.py

```diff
@@ -0,0 +1,803 @@
+        self.node_probabilities = counts / counts.sum()
+        self.node_probabilities_cumsum = np.cumsum(self.node_probabilities)
+        self.all_relations = all_relations
+        self.term_relations = term_relations
+        self.negatives_buffer = NegativesBuffer([])  # Buffer to store negative samples, to reduce calls to sampling method
+        self.negatives_buffer_size = 2000
+
+    def init_embeddings(self):
+        """Randomly initialize vectors for the items in the vocab."""
+        shape = (len(self.wv.index2word), self.size)
+        self.wv.syn0 = self.np_random.uniform(self.init_range[0], self.init_range[1], shape)
+
+    def get_candidate_negatives(self):
+        """
+        Returns candidate negatives of size `self.negative` from the negative examples buffer.
+
+        Returns
+        --------
+        numpy array
+            numpy array of shape (`self.negative`,) containing indices of negative nodes.
+        """
+
+        if self.negatives_buffer.is_empty():
+            # Note: np.random.choice much slower than random.sample for large populations, possible bottleneck
+            uniform_numbers = np.random.random(self.negatives_buffer_size)
+            cumsum_table_indices = np.searchsorted(self.node_probabilities_cumsum, uniform_numbers)
+            self.negatives_buffer = NegativesBuffer(cumsum_table_indices)
+        return self.negatives_buffer.get_items(self.negative)
+
+    @staticmethod
+    def has_duplicates(array):
+        """
+        Returns whether or not the input array has any duplicates.
+
+        Parameters
+        ----------
+        array : iterable of hashables
+            input array to checked, should contain hashable items.
+
+        Returns
+        -------
+        bool
+            Whether the input array contains any duplicates
+        """
+        seen = set()
+        for value in array:
+            if value in seen:
+                return True
+            seen.add(value)
+        return False
+
+    def sample_negatives(self, node_index):
+        """
+        Return a sample of negatives for the given node.
+
+        Parameters
+        ----------
+        node_index : int
+            index of the positive node for which negative samples are to be returned.
+
+        Returns
+        --------
+        numpy array
+            numpy array of shape (self.negative,) containing indices of negative nodes for the given node index.
+        """
+        node_relations = self.term_relations[node_index]
+        positive_fraction = len(node_relations) / len(self.term_relations)
+        if positive_fraction < 0.01:
+            # If number of positive relations is a small fraction of total nodes
+            # re-sample till no positively connected nodes are chosen
+            indices = self.get_candidate_negatives()
+            times_sampled = 1
+            while (set(indices) & node_relations) or self.has_duplicates(indices):
```

Comment:
Maybe change the order of conditions (if `indices` with duplicates, this will be faster), or this line doesn't affect performance?